File: agent/crates/agent-core/src/command_handler.rs
#![allow(unused_imports, unused_variables, dead_code, unused_mut)]
use edr_sdk::proto::fleet::{ServerCommand, server_command::Command};
use isolation::IsolationManager;
use osquery_client::OsqueryCollector;
use serde_json::Value;
use std::sync::Arc;
use tracing::{info, warn};

pub struct CommandHandler {
    pub osquery: Arc<OsqueryCollector>,
    pub isolation: IsolationManager,
}

impl CommandHandler {
    /// Processes a server command and returns a JSON status response.
    ///
    /// Handles different command types with specific actions:
    /// - `Isolate`: Isolates or de-isolates the process based on the flag.
    /// - `ConfigUpdate`: Acknowledges the configuration update.
    /// - `Ack`: Acknowledges the message.
    ///
    /// # Errors
    ///
    /// Returns an error if the command is missing from the message or if an isolation operation fails.
    ///
    /// # Examples
    ///
    /// ```
    /// # async fn example() {
    /// let handler = CommandHandler { /* ... */ };
    /// let cmd = ServerCommand { command: Some(Command::Ack(())) };
    /// let result = handler.handle(cmd).await;
    /// assert!(result.is_ok());
    /// # }
    /// ```
    pub async fn handle(&self, msg: ServerCommand) -> Result<Value, String> {
        let command = msg.command.ok_or("missing command")?;

        match command {
            Command::Isolate(iso) => {
                if iso.isolate {
                    self.isolation.isolate().await.map_err(|e| e.to_string())?;
                    Ok(serde_json::json!({"status": "isolated"}))
                } else {
                    self.isolation
                        .de_isolate()
                        .await
                        .map_err(|e| e.to_string())?;
                    Ok(serde_json::json!({"status": "unisolated"}))
                }
            }
            Command::ConfigUpdate(_cfg) => Ok(serde_json::json!({"status": "config_updated"})),
            Command::Ack(_) => Ok(serde_json::json!({"status": "acked"})),
        }
    }
}

File: agent/crates/agent-core/src/command_handler.py
import json
from typing import Any

from isolation import IsolationManager
from osquery-client import OsqueryCollector

from dataclasses import dataclass

@dataclass
class CommandHandler:
    osquery: OsqueryCollector
  isolation: IsolationManager

    async def handle(self, msg: dict[str, Any]) -> dict[str, str]:
        """
        Handle a command received from the Fleet server.
        """

        command = msg.get("command")

        if command is None:
            raise ValueError("missing command")

        command_type = command.get("type")

        if command_type == "isolate":
            isolate = command.get("isolate")

            if isolate:
                await self.isolation.isolate()

                return {
                    "status": "isolated"
                }

            else:
                await self.isolation.de_isolate()

                return {
                    "status": "unisolated"
                }

        elif command_type == "config_update":
            return {
                "status": "config_updated"
            }

        elif command_type == "ack":
            return {
                "status": "acked"
            }

        else:
            raise ValueError(
                f"unknown command type: {command_type}"
            )
